File: src/lexer/chars.rs
/// 文字アクセスメソッド群。
///
/// `Lexer` の現在読み取り位置（`pos`）周辺の文字を参照・消費する低レベルヘルパー。
/// すべてのメソッドは `scan.rs` で定義された `Lexer` に対する `impl` ブロックとして提供する。
use super::scan::Lexer;

impl Lexer {
    /// 現在位置の文字を返す（消費しない）。
    ///
    /// # 戻り値
    /// `Some(char)` または入力終端の場合 `None`
    pub(super) fn ch(&self) -> Option<char> {
        self.chars.get(self.pos).copied()
    }

    /// 現在位置の 1 つ先の文字を返す（消費しない）。
    ///
    /// # 戻り値
    /// `Some(char)` または範囲外の場合 `None`
    pub(super) fn ch1(&self) -> Option<char> {
        self.chars.get(self.pos + 1).copied()
    }

    /// 現在位置の 2 つ先の文字を返す（消費しない）。
    ///
    /// トリプルクォート終端の判定などで使用する。
    ///
    /// # 戻り値
    /// `Some(char)` または範囲外の場合 `None`
    pub(super) fn ch2(&self) -> Option<char> {
        self.chars.get(self.pos + 2).copied()
    }

    /// 現在位置の文字を返しつつ `pos` を 1 進める。
    ///
    /// # 戻り値
    /// `Some(char)` または入力終端の場合 `None`（`pos` は変化しない）
    pub(super) fn bump(&mut self) -> Option<char> {
        let ch = self.chars.get(self.pos).copied();
        if ch.is_some() {
            self.pos += 1;
        }
        ch
    }
}

File: impl_python/lexer/chars.py
# git SHA: 08f19f554735e8588bc1f4bd2e2b300b43e4a31a
"""Character access helpers for the Lexer (mirrors src/lexer/chars.rs)."""
from __future__ import annotations
from typing import Optional


class _LexerChars:
    """Mixin providing ch/ch1/ch2/bump character access."""

    _chars: list[str]
    _pos: int

    def _ch(self) -> Optional[str]:
        return self._chars[self._pos] if self._pos < len(self._chars) else None

    def _ch1(self) -> Optional[str]:
        p = self._pos + 1
        return self._chars[p] if p < len(self._chars) else None

    def _ch2(self) -> Optional[str]:
        p = self._pos + 2
        return self._chars[p] if p < len(self._chars) else None

    def _bump(self) -> Optional[str]:
        c = self._chars[self._pos] if self._pos < len(self._chars) else None
        if c is not None:
            self._pos += 1
        return c
